### PR TITLE
0.17: Backport #14453

### PR DESCRIPTION
**Pull request by promag**
Backport #14453 to 0.17

### CODE REVIEW

**Comment by promag** [comment]
In 0.17 branch `encryptwallet` stops the node, which changed in #12493. For that reason the test is adjusted.

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -2622,7 +2615,18 @@ static UniValue walletpassphrase(const JSONRPCRequest& request)
     pwallet->TopUpKeyPool();
 
     pwallet->nRelockTime = GetTime() + nSleepTime;
-    RPCRunLater(strprintf("lockwallet(%s)", pwallet->GetName()), std::bind(LockWallet, pwallet), nSleepTime);
+
+    // Keep a weak pointer to the wallet so that it is possible to unload the
+    // wallet before the following callback is called. If a valid shared pointer
+    // is acquired in the callback then the wallet is still loaded.
+    std::weak_ptr<CWallet> weak_wallet = wallet;
```
Comment (Empact) [inline src/wallet/rpcwallet.cpp]: nit: seems construction would be marginally better
https://en.cppreference.com/w/cpp/memory/weak_ptr/weak_ptr
https://en.cppreference.com/w/cpp/memory/weak_ptr/operator%3D

Comment (laanwj) [reply src/wallet/rpcwallet.cpp]: I don't think this is an appropriate comment for a backport as this change exists literally in the commit for master: https://github.com/bitcoin/bitcoin/pull/14453/commits/321decffa1fbf213462d97e5372bd0c4eeb99635

**Comment by laanwj** [comment]
utACK dcb032dcdf8e8300a84ae8ec6cfde0da5381f56f

**Comment by MarcoFalke** [comment]
utACK dcb032dcdf8e8300a84ae8ec6cfde0da5381f56f

**Pull request merged**